### PR TITLE
CBG-2052: Implement Blip Unsub changes for Connected Client protocol

### PR DESCRIPTION
**Pull request by IsaacLambat**
CBG-2052

-  Replaced changes feed terminator option with cancellable context
- Unsub changes cancels the context if it exists, otherwise returns an error
- Unsub changes uses a lock to make sure that the context is not created by subChanges while unsubChanges is checking if the context exists
- Added a lock for when closing the blip handler
- Added testing for this

- Changed changes options `Ctx` to be called `LoggingCtx`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/391

### CODE REVIEW

File: db/blip_handler.go [hunk 1]
```diff
@@ -263,8 +267,8 @@ func (bh *blipHandler) handleSubChanges(rq *blip.Message) error {
 }
 
 func (bh *blipHandler) handleUnsubChanges(rq *blip.Message) error {
-	return base.HTTPErrorf(http.StatusNotImplemented, "unsubChanges not implemented yet")
-	// TODO: Implement unsubChanges
+	bh.changesCtxCancel()
```
Comment (bbrks) [inline db/blip_handler.go:272]: What happens if a client calls `unsubChanges` before a `subChanges` has been issued? Does it panic? Can we handle that better if so?

File: rest/blip_api_test.go [hunk 2]
```diff
@@ -4492,3 +4492,43 @@ func TestBlipLegacyAttachDocUpdate(t *testing.T) {
...
+	unsubChangesRequest := blip.NewRequest()
+	unsubChangesRequest.SetProfile(db.MessageUnsubChanges)
+	err = btc.pullReplication.sendMsg(unsubChangesRequest)
+	require.NoError(t, err)
+	unsubChangesResponse := unsubChangesRequest.Response()
+	_, err = unsubChangesResponse.Body()
+	require.NoError(t, err)
+
+	// Confirm no more changes being sent
+	resp = rt.updateDoc("doc2", "", `{"key":"val1"}`)
```
Comment (bbrks) [inline rest/blip_api_test.go]: This test might be racy given it's not a synchronous operation to call `unsubChanges`. If the changes feed takes a while to actually exit, doc2 might end up being sent.

I'm not exactly sure of the fix here - the replication protocol wording implies unsub is async, but it's possible we can block inside the test on stats instead to ensure timing.

**Review by bbrks — CHANGES_REQUESTED** [review]
A couple of comments related to edge cases/potential race condition in the test.

File: db/changes.go [hunk 3]
```diff
@@ -1305,7 +1306,12 @@ func generateBlipSyncChanges(database *Database, inChannels base.Set, options Ch
 
 	// Store one-shot here to protect
 	isOneShot := !options.Continuous
-	err, forceClose = GenerateChanges(context.Background(), database, inChannels, options, docIDFilter, send)
+	// Use the context if provided in the options
+	ctx := options.ChangesCtx
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	err, forceClose = GenerateChanges(ctx, database, inChannels, options, docIDFilter, send)
```
Comment (adamcfraser) [inline db/changes.go]: I don't think closing the context being passed to GenerateChanges is sufficient to terminate all goroutines associated with a changes feed.  Closing the changes terminator channel is the only thing that does that today - you can search for usages of that terminator to see where it's used - only one of those also does a check for the context being cancelled.  We could refactor changes to use a context throughout instead of the terminator channel, but otherwise I think we need to continue using the terminator to stop changes.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/blip_api_test.go [hunk 4]
```diff
@@ -4494,3 +4494,49 @@ func TestBlipLegacyAttachDocUpdate(t *testing.T) {
...
+	require.EqualValues(t, 1, activeReplStat.Value())
+
+	// Unsub changes
+	unsubChangesRequest := blip.NewRequest()
+	unsubChangesRequest.SetProfile(db.MessageUnsubChanges)
+	err = btc.pullReplication.sendMsg(unsubChangesRequest)
+	require.NoError(t, err)
+	unsubChangesResponse := unsubChangesRequest.Response()
+	_, err = unsubChangesResponse.Body()
+	require.NoError(t, err)
```
Comment (bbrks) [inline rest/blip_api_test.go]: Doing this doesn't really match how the `BlipTesterClient` works - which is a higher level API that hides a lot of the BLIP/replication messaging.

You could just move this code into a method on `BlipTesterClient`, so you can call `require.NoError(t, btc.UnsubChanges())`, in the same way you call `StartPull()` to start them.

File: rest/blip_api_test.go [hunk 5]
```diff
@@ -4494,3 +4494,49 @@ func TestBlipLegacyAttachDocUpdate(t *testing.T) {
...
+func TestUnsubChanges(t *testing.T) {
+	base.SetUpTestLogging(t, base.LevelInfo, base.KeyAll)
+	rt := NewRestTester(t, &RestTesterConfig{guestEnabled: true})
+	defer rt.Close()
+
+	btc, err := NewBlipTesterClientOptsWithRT(t, rt, nil)
+	require.NoError(t, err)
+	defer btc.Close()
+
+	// Sub changes
```
Comment (bbrks) [inline rest/blip_api_test.go:4627]: I think I mentioned in previous comment about making sure it works - but would be nice to bake this into a test assertion:

Can we run an `UnsubChanges` before the first `SubChanges` to make sure this doesn't panic/cause other issues?

File: db/blip_handler.go [hunk 6]
```diff
@@ -263,8 +268,14 @@ func (bh *blipHandler) handleSubChanges(rq *blip.Message) error {
 }
 
 func (bh *blipHandler) handleUnsubChanges(rq *blip.Message) error {
-	return base.HTTPErrorf(http.StatusNotImplemented, "unsubChanges not implemented yet")
-	// TODO: Implement unsubChanges
+	bh.lock.Lock()
+	defer bh.lock.Unlock()
```
Comment (bbrks) [inline db/blip_handler.go]: Note to self: Take a closer look at what this lock is guarding

File: db/blip_handler.go [hunk 7]
```diff
@@ -198,6 +198,11 @@ func (bh *blipHandler) handleSubChanges(rq *blip.Message) error {
 		return fmt.Errorf("blipHandler already has an outstanding continous subChanges.  Cannot open another one")
 	}
 
+	// Create ctx if it does not exist or has been closed
+	if bh.changesCtx == nil || bh.changesCtx.Err() != nil {
```
Comment (bbrks) [inline db/blip_handler.go]: I think it might actually be safer for us _not_ to handle `nil` contexts - it's not recommended to pass around nil contexts, and we want to find the code paths that are doing so.

As far as I can tell, this should never be nil, which would save a fair amount of awkward nil checking for `changesCtx`/`changesCtxCancel`.

File: db/changes.go [hunk 8]
```diff
@@ -524,8 +524,11 @@ func (db *Database) MultiChangesFeed(chans base.Set, options ChangesOptions) (<-
 		return nil, nil
 	}
 
-	if (options.Continuous || options.Wait) && options.Terminator == nil {
-		base.WarnfCtx(db.Ctx, "MultiChangesFeed: Terminator missing for Continuous/Wait mode")
+	if options.ChangesCtx == nil {
+		if options.Continuous || options.Wait {
+			base.WarnfCtx(db.Ctx, "MultiChangesFeed: Changes Context missing for Continuous/Wait mode")
+		}
+		options.ChangesCtx = context.Background()
```
Comment (bbrks) [inline db/changes.go]: I'm wondering if this is a situation where we _want_ to find out about this (at dev time) by returning an error/panicing, rather than warning and carrying on with no way to kill the changes goroutines that are started up.

Looks like this is ancient code [if you step back far enough](https://github.com/couchbase/sync_gateway/commit/6d381ffea63ffdbf2efbfe0337ac68de95255334) and I've neither seen this warning anywhere before, nor do I think it's a risk for us to start to be stricter about this.

**Review by bbrks — CHANGES_REQUESTED** [review]
A few comments but I do still need to take a closer look at the locking on `blipHandler` to figure out what its intended use is.

File: rest/blip_api_test.go (hunk 5 above)
Comment (IsaacLambat) [reply rest/blip_api_test.go:4627]: I did try this, and it did panic initially but it is fixed now. The problem with having it on a test assertion is that Blip recovers from the panic and just logs it instead, so not sure how I could assert on this unless I assert on the logs containing the panic text.

Will try to come up with something.

File: db/changes.go (hunk 8 above)
Comment (IsaacLambat) [reply db/changes.go]: The main point of adding this context is because lots of tests call the changes feed directly, bypassing the context initialisation. I will remove this changes context check, and change lots of tests to have the context added in the changes options.

File: rest/blip_api_test.go (hunk 5 above)
Comment (bbrks) [reply rest/blip_api_test.go:4627]: BLIP should return an error code and a message in the body if the handler paniced

File: db/changes.go (hunk 8 above)
Comment (bbrks) [reply db/changes.go]: Understood. I think it makes sense to force the tests to pass in a context, otherwise we'd potentially never stop changes feeds as we have a non-cancellable background context in use.

File: db/blip_handler.go (hunk 6 above)
Comment (bbrks) [reply db/blip_handler.go]: As discussed - rename this existing lock to indicate it's guarding `allowedAttachments` - and create a new `changesCtxLock` for your usage.

File: db/changes.go [hunk 9]
```diff
@@ -400,7 +400,7 @@ func (db *Database) changesFeed(singleChannelCache SingleChannelCache, options C
 				paginationOptions.Limit = base.MinInt(remainingLimit, queryLimit)
 			}
 
-			// TODO: pass db.Ctx down to changeCache?
+			// TODO: pass db.LoggingCtx down to changeCache?
```
Comment (bbrks) [inline db/changes.go]: unintentional change

```suggestion
			// TODO: pass db.Ctx down to changeCache?
```

**Review by bbrks — CHANGES_REQUESTED** [review]

File: db/blip_handler.go [hunk 10]
```diff
@@ -264,8 +266,14 @@ func (bh *blipHandler) handleSubChanges(rq *blip.Message) error {
...
 
 func (bh *blipHandler) handleUnsubChanges(rq *blip.Message) error {
-	return base.HTTPErrorf(http.StatusNotImplemented, "unsubChanges not implemented yet")
-	// TODO: Implement unsubChanges
+	bh.changesCtxLock.Lock()
+	defer bh.changesCtxLock.Unlock()
+
+	if !bh.activeSubChanges.IsTrue() {
+		return base.HTTPErrorf(http.StatusBadRequest, "No subChanges subscription active to unsubscribe from")
+	}
```
Comment (bbrks) [inline db/blip_handler.go]: I think we can probably get away without doing this check here. We may still want to try and cancel the context if we have one, even if activeSubChanges is reporting false for whatever reason.

```suggestion
```

**Review by bbrks — CHANGES_REQUESTED** [review]
One final small change on UnsubChanges then I think this is good to go

**Review by bbrks — APPROVED** [review]
LGTM - pending tests

**Review by adamcfraser — APPROVED** [review]
